cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java

```diff
@@ -100,26 +73,78 @@ public void run(JavaSparkExecutionContext sec) throws Exception {
+      runWithCheckpointing(pipelineSpec, pipelinePhase, sec);
+    jssc.start();
     boolean stopped = false;
     try {
       // most programs will just keep running forever.
       // however, when CDAP stops the program, we get an interrupted exception.
       // at that point, we need to call stop on jssc, otherwise the program will hang and never stop.
-      stopped = streamingContext.awaitTerminationOrTimeout(Long.MAX_VALUE);
+      stopped = jssc.awaitTerminationOrTimeout(Long.MAX_VALUE);
     } finally {
       if (!stopped) {
-        streamingContext.stop(true, true);
+        jssc.stop(true, true);
       }
     }
   }
+
+  private JavaStreamingContext runWithoutCheckpointing(final DataStreamsPipelineSpec pipelineSpec,
+                                                       final PipelinePhase pipelinePhase,
+                                                       final JavaSparkExecutionContext sec) throws Exception {
+    JavaStreamingContext jssc = new JavaStreamingContext(
+      new JavaSparkContext(), Durations.milliseconds(pipelineSpec.getBatchIntervalMillis()));
+    SparkStreamingPipelineRunner runner = new SparkStreamingPipelineRunner(sec, jssc, true);
+    // TODO: figure out how to get partitions to use for aggregators and joiners.
+    // Seems like they should be set at configure time instead of runtime? but that requires an API change.
+    runner.runPipeline(pipelinePhase, StreamingSource.PLUGIN_TYPE, sec, new HashMap<String, Integer>());
+    return jssc;
+  }
+
+  private JavaStreamingContext runWithCheckpointing(final DataStreamsPipelineSpec pipelineSpec,
+                                                    final PipelinePhase pipelinePhase,
+                                                    final JavaSparkExecutionContext sec) throws Exception {
+    // Get the location of the checkpoint directory.
+    String pipelineName = sec.getApplicationSpecification().getName();
+    String pipelineId = sec.getSpecification().getProperty(DataStreamsSparkLauncher.PIPELINE_UUID);
+
+    // there isn't any way to instantiate the fileset except in a TxRunnable, so need to use a reference.
+    final AtomicReference<Location> checkpointBaseRef = new AtomicReference<>();
+    sec.execute(new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        FileSet checkpointFileSet = context.getDataset(DataStreamsApp.CHECKPOINT_FILESET);
+        checkpointBaseRef.set(checkpointFileSet.getBaseLocation());
+      }
+    });
+    Location pipelineCheckpointDir = checkpointBaseRef.get().append(pipelineName).append(pipelineId);
+    final String checkpointDir = pipelineCheckpointDir.toURI().toString();
+
+    JavaStreamingContextFactory contextFactory = new JavaStreamingContextFactory() {
+      @Override
+      public JavaStreamingContext create() {
+        JavaStreamingContext jssc = new JavaStreamingContext(
```

Comment:
This is basically the same as the `runWithoutCheckpointing`, except the call to `jssc.checkpoint`, right? Better have a method `run(@Nullable Location checkoutLocation)`.